drivers/imx6-enet.c

```diff
@@ -793,6 +819,135 @@ static void enet_setLinkState(void *arg, int state)
 	}
 }
 
+
+#define _TP_ETHTYPE "\x05\xDD" /* eth frame type 0x05DD is undefined */
+#define _TP_10DIG   "0123456789"
+#define TEST_PACKET "ddddddssssss" _TP_ETHTYPE \
+	_TP_10DIG _TP_10DIG _TP_10DIG _TP_10DIG _TP_10DIG _TP_10DIG _TP_10DIG
+#define TEST_PACKET_LEN (sizeof((TEST_PACKET)) - 1)
+
+
+/* self-test RX input function */
+static err_t test_netif_input(struct pbuf *p, struct netif *netif)
+{
+	uint8_t buf[TEST_PACKET_LEN]; /* used only if pbuf is fragmented (should not happen) */
+	enet_priv_t *priv = netif->state;
+
+	bool is_valid_pkt = true;
+
+	/* verify contents */
+	if ((p->len != TEST_PACKET_LEN + ETH_PAD_SIZE)) {
+#ifdef ENET_VERBOSE
+		enet_printf(priv, "self-test RX: invalid packet length");
+#endif
+		is_valid_pkt = false;
+	}
+	uint8_t *data = pbuf_get_contiguous(p, buf, sizeof(buf), TEST_PACKET_LEN, ETH_PAD_SIZE);
+	if (data == NULL || memcmp(TEST_PACKET, data, TEST_PACKET_LEN) != 0) {
+#ifdef ENET_VERBOSE
+		if (data == NULL) {
+			data = p->payload;
+		}
+		enet_printf(priv, "self-test RX: invalid packet contents");
+		enet_printf(priv, "0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x",
+			data[0], data[1], data[2], data[3],
+			data[4], data[5], data[6], data[7]);
+#endif
+		is_valid_pkt = false;
+	}
+	pbuf_free(p);
+
+	mutexLock(priv->selfTest.rx_lock);
+	if (is_valid_pkt) {
+		priv->selfTest.rx_cnt += 1;
+	}
+	mutexUnlock(priv->selfTest.rx_lock);
+	condBroadcast(priv->selfTest.rx_cond);
```

Comment:
Call before unlocking mutex?

Reply:
Yeah, the lock/cond pattern should probably be done as you've described but it turns out that it's slower on our OS (cond broadcast would result in immediate reschedule).

Further investigation proved it's not easy (especially in SMP) to optimize this code path and it usually isn't in various pthread implementations.

Reply:
OK, in that case you don't need to lock the mutex if `is_valid_pkt` is false.

Reply:
renamed `rx_cnt` to `rx_valid` and explicitly mark receiving invalid packet to break `condWait` earlier
(testing showed it's common to receive broken packet when PHY configuration is invalid)